### PR TITLE
feat(stdlib): Add lookupCI for assocarray and Node

### LINKED ISSUE #629: lookupCI() is not supported
`Error: 'lookupCI' is not a function and cannot be called.`

From [Roku SDK](https://developer.roku.com/en-gb/docs/references/brightscript/interfaces/ifassociativearray.md): 
_"LookupCI(key as String) as Dynamic
Description
Same as the Lookup() method except that the key comparison is always case insensitive, regardless of the case mode."_

### PR DESCRIPTION
**Pull request by Vasya-M**
solution $629 issue
what were done:
- fixed `CreateObject("roSGNode", "node")`
- added support for `lookupCI` for AA and Nodes
- added UT and e2e tests

resolves #629  

UPD:
 - Fixed logic of using/saving keys in AA (saving last key used by `[]`, or lowercase key if accessed by `.`)

### CODE REVIEW

File: src/brsTypes/components/ComponentFactory.ts [hunk 1]
```diff
@@ -18,6 +18,7 @@ import {
 
 export enum BrsComponentName {
     Node = "Node",
+    Node_lowerCase = "node",
```
Comment (lkipke) [inline src/brsTypes/components/ComponentFactory.ts:21]: Is this a problem with _all_ components? What if I did `createObject("roSGNode", "group")` -- I'm assuming I'd encounter the same issue.

Also I think we should probably do this in a separate PR, just because it's not strictly related to `lookupCI`.

**Review by lkipke — CHANGES_REQUESTED** [review]
I think the `lookupCI` code looks great, but let's pull out the case-sensitive `createObject` code into its own PR 🙂

File: src/brsTypes/components/ComponentFactory.ts (hunk 1 above)
Comment (lkipke) [reply src/brsTypes/components/ComponentFactory.ts:21]: Also, what about cases like `createObject("RoSGNode", "nOdE")`? Should those also create a Node object?

Comment (Vasya-M) [reply src/brsTypes/components/ComponentFactory.ts:21]: yep, you are right, all naming for type of node is case insensitive (works for user components too).

Comment (Vasya-M) [reply src/brsTypes/components/ComponentFactory.ts]: reverted fix for #638 issue

**Review by lkipke — APPROVED** [review]
This looks great, thanks for the updates!

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 2]
```diff
@@ -235,6 +239,17 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
         },
     });
 
+    private lookupCI = new Callable("lookupCI", {
```
Comment (alimnios72) [inline src/brsTypes/components/RoAssociativeArray.ts:242]: Nit, can you add docs to this function please?

File: src/brsTypes/components/RoSGNode.ts [hunk 3]
```diff
@@ -801,6 +802,8 @@ export class RoSGNode extends BrsComponent implements BrsValue, BrsIterable {
         },
     });
 
+    private lookupCI = new Callable("lookupCI", this.lookup.signatures[0]);
```
Comment (alimnios72) [inline src/brsTypes/components/RoSGNode.ts:805]: please add docs as well. Also, is this function expected to do the same as `lookup` or what is the difference between the 2?

File: test/brsTypes/components/RoAssociativeArray.test.js [hunk 4]
```diff
@@ -372,6 +372,20 @@ describe("RoAssociativeArray", () => {
 
                 result2 = aa.get(new BrsString("KeY2"));
                 expect(result2).toBe(v2);
+
+                // check lookupCI method
+                let lookupCI = aa.getMethod("lookupCI");
```
Comment (alimnios72) [inline test/brsTypes/components/RoAssociativeArray.test.js:377]: can you please add a new test suite for these test cases? Following the same template as we did for other functions, i.e.:
```
describe("lookupCI", () => {
...
}
```
Also it would be interesting to test this function when `SetModeCaseSensitive` is set and when it's not

**Review by alimnios72 — COMMENTED** [review]
overall looks good, a few comments below

File: test/brsTypes/components/RoAssociativeArray.test.js (hunk 4 above)
Comment (Vasya-M) [reply test/brsTypes/components/RoAssociativeArray.test.js:377]: ok, but this method is always is case insensitive, so there is no need test it under different states of `SetModeCaseSensitive`, just need to test if CaseSensitive is enabled
just now tested it and noticed that there is bug
```
brs> aa = {}
brs> aa.SetModeCaseSensitive()
brs> aa["AA"] = 0
brs> ?aa.aa
invalid ' will be 0 in roku
brs> ?aa.lookupCI("aa")
invalid ' will be 0 in roku
brs> ?aa
<Component: roAssociativeArray> =
{
    AA: 0
}

```

File: src/brsTypes/components/RoSGNode.ts (hunk 3 above)
Comment (alimnios72) [reply src/brsTypes/components/RoSGNode.ts:805]: You had one job brightscript, LOL 😆

File: test/brsTypes/components/RoAssociativeArray.test.js (hunk 4 above)
Comment (alimnios72) [reply test/brsTypes/components/RoAssociativeArray.test.js:377]: Thanks for checking, maybe for posterity is worth having its own different test case hoping that one day Roku fixes the bug and we can update the tests appropriately

File: src/brsTypes/components/RoSGNode.ts [hunk 5]
```diff
@@ -435,30 +436,53 @@ export class RoSGNode extends BrsComponent implements BrsValue, BrsIterable {
...
-        let fieldType = kind || FieldKind.fromBrsType(value);
         let field = this.fields.get(mapKey);
 
         if (!field) {
-            // RBI does not create a new field if the value isn't valid.
-            if (fieldType) {
-                field = new Field(value, fieldType, alwaysNotify);
-                this.fields.set(mapKey, field);
-            }
+            // TODO: change to using interpreter.stderr
```
Comment (alimnios72) [inline src/brsTypes/components/RoSGNode.ts:448]: If having this creates issues int the current PR (which I think it's breaking unit tests right now) I'm ok with opening a github issue and tackling this work in a separate PR

Comment (Vasya-M) [reply src/brsTypes/components/RoSGNode.ts]: yep, did it

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 6]
```diff
@@ -93,18 +92,49 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
         // method with the desired name separately? That last bit would work but it's pretty gross.
         // That'd allow roArrays to have methods with the methods not accessible via `arr["count"]`.
         // Same with RoAssociativeArrays I guess.
-        let indexValue = this.modeCaseSensitive ? index.value : index.value.toLowerCase();
-        return this.elements.get(indexValue) || this.getMethod(index.value) || BrsInvalid.Instance;
+        return (
+            this.findElement(index.value, isCaseSensitiveGet) ||
```
Comment (alimnios72) [inline src/brsTypes/components/RoAssociativeArray.ts:96]: I think you can get rid of `findElement` function and just do the following:
```
this.elements.get(this.findElementKey(index.value), isCaseSensitiveGet)
```
because if the key is `undefined` the map should return that as well

**Review by alimnios72 — COMMENTED** [review]
A few last comments below

File: src/brsTypes/components/RoAssociativeArray.ts (hunk 6 above)
Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts:96]: image
looks like no, it can not accept `undefined`

Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts:96]: image

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 7]
```diff
@@ -93,18 +92,49 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
...
+        let oldKey = this.findElementKey(index.value);
+        if (!this.modeCaseSensitive && oldKey) {
+            this.elements.delete(oldKey);
+        }
+
+        let indexValue = isCaseSensitiveSet ? index.value : index.value.toLowerCase();
         this.elements.set(indexValue, value);
         return BrsInvalid.Instance;
     }
+    /** if AA is in insensitive mode, it means that we should do insensitive search of real key */
```
[the remark this answers is not in the document]
Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts:117]: done

File: src/brsTypes/components/BrsComponent.ts [hunk 8]
```diff
@@ -58,9 +58,10 @@ export interface BrsIterable {
     /**
      * Retrieves an element from this component at the provided `index`.
      * @param index the index in this component from which to retrieve the desired element.
+     * @param isCaseSensitiveGet determinate whether operation of getting should be case sensitive or not.
      * @returns the element at `index` if one exists, otherwise throws an Error.
      */
-    get(index: BrsType): BrsType;
+    get(index: BrsType, isCaseSensitiveGet?: boolean): BrsType;
```
[the remark this answers is not in the document]
Comment (Vasya-M) [reply src/brsTypes/components/BrsComponent.ts]: done

File: test/brsTypes/components/RoAssociativeArray.test.js [hunk 9]
```diff
@@ -374,5 +374,40 @@ describe("RoAssociativeArray", () => {
                 expect(result2).toBe(v2);
             });
         });
+        describe("lookupCI", () => {
```
[the remark this answers is not in the document]
Comment (Vasya-M) [reply test/brsTypes/components/RoAssociativeArray.test.js:378]: done

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 10]
```diff
@@ -77,7 +76,7 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
             .map((value: BrsType) => value);
     }
 
-    get(index: BrsType) {
+    get(index: BrsType, isCaseSensitiveGet = false) {
```
[the remark this answers is not in the document]
Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts]: done

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 11]
```diff
@@ -93,19 +92,51 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
...
+            return elementKey;
+        }
+
+        let realElementKey = undefined;
+        for (let key of this.elements.keys()) {
+            if ((useCaseSensitiveSearch ? key : key.toLowerCase()) === elementKey) {
+                realElementKey = key;
+                break;
+            }
+        }
```
Comment (sjbarag) [inline src/brsTypes/components/RoAssociativeArray.ts:131]: This seems less than ideal, as it'll cause an `O(n)` lookup for every access once `setCaseInsensitive` is called.  I'm curious if there's an alternative solution we can take that won't require us to iterate across every element every time.  Could we maintain a case-insensitive `Map` next to the case-sensitive one, and write to both for every `set()`? It'd of course require twice as much memory, but would maintain constant time lookups.

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 12]
```diff
@@ -223,8 +258,22 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
...
+    /** Given a key, returns the value associated with that key.
+     * This method is case insensitive either-or case sensitive, depends on whether `setModeCasesensitive` was called or not.
+     */
     private lookup = new Callable("lookup", {
+        signature: {
+            args: [new StdlibArgument("key", ValueKind.String)],
+            returns: ValueKind.Dynamic,
+        },
+        impl: (interpreter: Interpreter, key: BrsString) => {
+            let lKey = key.value;
```
Comment (sjbarag) [inline src/brsTypes/components/RoAssociativeArray.ts:270]: Perhaps I've missed something — why's this variable `lKey` while the rest are `key`?

**Review by sjbarag — CHANGES_REQUESTED** [review]
This is a good start, @Vasya-M!  I'm a bit worried by the `O(n)` lookups this introduces in case-insensitive mode, so let's see what we can do to keep that from being an exhaustive search 😃

File: src/brsTypes/components/RoAssociativeArray.ts (hunk 12 above)
Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts:270]: idk why, but it was as is before my changes

File: src/brsTypes/components/RoAssociativeArray.ts (hunk 11 above)
Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts:131]: Hey, I even forget about it, a great catch. 
So I've taken your idea of using a second map and used it for mapping lower case keys to case sensitive keys. so it will not consume so much memory, will not introduce additional computation and almost saves previous performance. 

but there is still be `O(n)` when we delete keys in case sensitive mode, cuz we can delete key that was used as alias in `KeysMap`

File: src/brsTypes/components/RoAssociativeArray.ts (hunk 12 above)
Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts]: fixed

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 13]
```diff
@@ -17,13 +17,15 @@ export interface AAMember {
 export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIterable {
     readonly kind = ValueKind.Object;
     elements = new Map<string, BrsType>();
+    /** keyMap - maps lowercased key to original one which used in this.elements to store value
+     * Main benefit of it is fast case insensitive access
+     */
```
Comment (lkipke) [inline src/brsTypes/components/RoAssociativeArray.ts:22]: nit: a little syntax
```suggestion
    /** Maps lowercased element name to original name used in this.elements.
     * Main benefit of it is fast, case-insensitive access.
     */
```

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 14]
```diff
@@ -142,7 +187,7 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
             returns: ValueKind.Void,
         },
         impl: (interpreter: Interpreter, key: BrsString, value: BrsType) => {
-            this.set(key, value);
+            this.set(key, value, true);
```
Comment (lkipke) [inline src/brsTypes/components/RoAssociativeArray.ts:190]: can we add a comment to make it clear what this boolean refers to?
```suggestion
            this.set(key, value, /* isCaseSensitive */ true);
```

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 15]
```diff
@@ -125,7 +154,23 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
...
         impl: (interpreter: Interpreter, str: BrsString) => {
-            let deleted = this.elements.delete(str.value.toLowerCase());
+            let key = this.findElementKey(str.value, this.modeCaseSensitive);
+            let deleted = key ? this.elements.delete(key) : false;
+
+            let lKey = str.value.toLowerCase();
+            if (this.keyMap.get(lKey) === key) {
+                this.keyMap.delete(lKey);
+                // if we have {"DD": 0, "dD": 0},  keyMap["dd"] is pointed to "DD",
+                // and we delets it, then we should find another value for case insensitive accessing ("dD")
```
Comment (lkipke) [inline src/brsTypes/components/RoAssociativeArray.ts:164]:
```suggestion
                // and we delete it, then we should find another value for case insensitive accessing ("dD")
```

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 16]
```diff
@@ -182,7 +228,9 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
                 return BrsInvalid.Instance;
             }
 
-            this.elements = new Map<string, BrsType>([...this.elements, ...obj.elements]);
+            obj.elements.forEach((value, key) => {
+                this.set(new BrsString(key), value, true);
+            });
```
Comment (lkipke) [inline src/brsTypes/components/RoAssociativeArray.ts:233]: what is this logic change needed for?

File: test/e2e/resources/components/roAssociativeArray.brs [hunk 17]
```diff
@@ -11,6 +11,7 @@ sub main()
     print "can delete elements: " aa.delete("baz")              ' => true
     print "can look up elements: " aa.lookup("foo") = "foo"     ' => true
     print "can look up elements (brackets): " aa["foo"] = "foo" ' => true
+    print "can case insensitive look up elements: " aa.lookupCI("foO") = "foo" ' => true
```
Comment (lkipke) [inline test/e2e/resources/components/roAssociativeArray.brs:14]: Can we add another call to `lookupCI` after we call `setModeCaseSensitive` as well?

**Review by lkipke — APPROVED** [review]
Overall, looks great! Thanks for all the hard work here!

File: src/brsTypes/components/RoAssociativeArray.ts (hunk 16 above)
Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts:233]: cuz with a new approach, we need add lowercase key to map too, and save original key, what `this.set` already do

**Review by alimnios72 — APPROVED** [review]
Other than the already given feedback this looks good to me

File: test/e2e/resources/components/roAssociativeArray.brs (hunk 17 above)
Comment (Vasya-M) [reply test/e2e/resources/components/roAssociativeArray.brs:14]: done

File: src/brsTypes/components/RoAssociativeArray.ts [hunk 18]
```diff
@@ -125,7 +154,23 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
...
         impl: (interpreter: Interpreter, str: BrsString) => {
-            let deleted = this.elements.delete(str.value.toLowerCase());
+            let key = this.findElementKey(str.value, this.modeCaseSensitive);
+            let deleted = key ? this.elements.delete(key) : false;
+
+            let lKey = str.value.toLowerCase();
+            if (this.keyMap.get(lKey) === key) {
+                this.keyMap.delete(lKey);
+                // if we have {"DD": 0, "dD": 0},  keyMap["dd"] is pointed to "DD",
+                // and we delete it, then we should find another value for case insensitive accessing ("dD")
```
Comment (sjbarag) [inline src/brsTypes/components/RoAssociativeArray.ts:164]: Thanks for calling this out!  This is an interesting case, and feels like `keyMap` should be `Map<string, Set<string>>`: a map of lowercase keys to all encountered cases of that key.  In your example with `{ "DD": 0, "dD": 1}`, `keyMap.get("dd")` would be the set `("DD", "dD")`.  When `someAA.delete("DD")` is called from BrightScript, this function removes the first element from that set (something like `let newSet = new Set([ ...keyMap.get(lKey) ].slice(1));`) in case-sensitive mode, and removes the entire entry (`this.keyMap.delete(lKey)`) in case-insensitive mode.

Lookups become pretty similar, using the element at index zero of this `Set<string>`.

**Review by sjbarag — CHANGES_REQUESTED** [review]
We're really close!  I've got one last idea to avoid an `O(n)` operation for every `.delete()` call, but otherwise this is looking really good :)  Thanks for sticking with it, @Vasya-M!

File: src/brsTypes/components/RoAssociativeArray.ts (hunk 18 above)
Comment (Vasya-M) [reply src/brsTypes/components/RoAssociativeArray.ts]: done

**Review by sjbarag — APPROVED** [review]
Looks great!  Thanks for the changes @Vasya-M 😃
